Can't build with OCaml 4.07.1 Win32

Here is the full log:
```
$ dune build -p stdlib-shims --verbose
# Workspace root: "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\.opam-switch\\build\\stdlib-shims.0.1.0"
# # Workspace root: "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\.opam-switch\\build\\stdlib-shims.0.1.0"
# Auto-detected concurrency: 4
Running[0]: C:\Users\dbely\.opam\4.07.1+local+vc15+x64\bin\ocamlc.opt.exe -config > C:\Users\dbely\AppData\Local\Temp\dune15f338.output
# # # Workspace root: "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\.opam-switch\\build\\stdlib-shims.0.1.0"
# # Auto-detected concurrency: 4
# Dune context:
#  {name = "default";
#    kind = "default";
#    profile = "release";
#    merlin = true;
#    for_host = None;
#    build_dir = "default";
#    toplevel_path =
#      Some External "C:/Users/dbely/.opam/4.07.1+local+vc15+x64/lib/toplevel";
#    ocaml_bin = External "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin";
#    ocaml =
#      External
#        "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocaml.exe";
#    ocamlc =
#      External
#        "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocamlc.opt.exe";
#    ocamlopt =
#      Some
#        External
#          "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocamlopt.opt.exe";
#    ocamldep =
#      External
#        "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocamldep.opt.exe";
#    ocamlmklib =
#      External
#        "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocamlmklib.opt.exe";
#    env =
#      map {"DUNE_CONFIGURATOR" :
#           "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocamlc.opt.exe";
#      "INSIDE_DUNE" : "1";
#      "MANPATH" :
#      "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\.opam-switch\\build\\stdlib-shims.0.1.0\\_build/install/default/bin";
#      "OCAMLFIND_IGNORE_DUPS_IN" :
#      "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\.opam-switch\\build\\stdlib-shims.0.1.0\\_build/install/default/lib";
#      "OCAMLPATH" :
#      "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\.opam-switch\\build\\stdlib-shims.0.1.0\\_build/install/default/lib"};
#    findlib_path =
#      [External "C:\\\\Users\\\\dbely\\\\.opam\\\\4.07.1+local+vc15+x64\\lib"];
#    arch_sixtyfour = true;
#    natdynlink_supported = true;
#    supports_shared_libraries = true;
#    opam_vars = map {};
#    ocaml_config =
#      {version = "4.07.1";
#        standard_library_default =
#          "C:/Users/dbely/.opam/4.07.1+local+vc15+x64/lib/ocaml";
#        standard_library =
#          "C:/Users/dbely/.opam/4.07.1+local+vc15+x64/lib/ocaml";
#        standard_runtime = "ocamlrun";
#        ccomp_type = "msvc";
#        c_compiler = "cl";
#        ocamlc_cflags = ["-nologo"; "-O2"; "-Gy-"; "-MD"];
#        ocamlopt_cflags = ["-nologo"; "-O2"; "-Gy-"; "-MD"];
#        bytecomp_c_compiler =
#          ["cl"; "-nologo"; "-O2"; "-Gy-"; "-MD";
#          "-D_CRT_SECURE_NO_DEPRECATE"];
#        bytecomp_c_libraries = ["advapi32.lib"; "ws2_32.lib"; "version.lib"];
#        native_c_compiler =
#          ["cl"; "-nologo"; "-O2"; "-Gy-"; "-MD";
#          "-D_CRT_SECURE_NO_DEPRECATE"];
#        native_c_libraries = ["advapi32.lib"; "ws2_32.lib"; "version.lib"];
#        cc_profile = [];
#        architecture = "amd64";
#        model = "default";
#        int_size = 63;
#        word_size = 64;
#        system = "win64";
#        asm = ["ml64"; "-nologo"; "-Cp"; "-c"; "-Fo"];
#        asm_cfi_supported = false;
#        with_frame_pointers = false;
#        ext_exe = ".exe";
#        ext_obj = ".obj";
#        ext_asm = ".asm";
#        ext_lib = ".lib";
#        ext_dll = ".dll";
#        os_type = "Win32";
#        default_executable_name = "camlprog.exe";
#        systhread_supported = true;
#        host = "msvc64";
#        target = "msvc64";
#        profiling = false;
#        flambda = false;
#        spacetime = false;
#        safe_string = false;
#        exec_magic_number = "Caml1999X023";
#        cmi_magic_number = "Caml1999I024";
#        cmo_magic_number = "Caml1999O023";
#        cma_magic_number = "Caml1999A023";
#        cmx_magic_number = "Caml1999Y023";
#        cmxa_magic_number = "Caml1999Z023";
#        ast_impl_magic_number = "Caml1999M023";
#        ast_intf_magic_number = "Caml1999N023";
#        cmxs_magic_number = "Caml1999D023";
#        cmt_magic_number = "Caml1999T024";
#        natdynlink_supported = true;
#        supports_shared_libraries = true;
#        windows_unicode = true};
#    which =
#      map {"ocaml" :
#           Some
#             External
#               "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocaml.exe";
#      "ocamlc" :
#      Some
#        External
#          "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocamlc.opt.exe";
#      "ocamlobjinfo" :
#      Some
#        External
#          "C:\\Users\\dbely\\.opam\\4.07.1+local+vc15+x64\\bin\\ocamlobjinfo.opt.exe"}}
Running[1]: (cd src && C:\Users\dbely\.opam\4.07.1+local+vc15+x64\bin\ocaml.exe -I +compiler-libs C:\Users\dbely\.opam\4.07.1+local+vc15+x64\.opam-switch\build\stdlib-shims.0.1.0\_build/.dune/default/src/dune.ml)
# Actual targets:
# - alias @@./default
Running[2]: (cd _build/default && C:\Users\dbely\.opam\4.07.1+local+vc15+x64\bin\ocamlopt.opt.exe -w -40 -g -a -o src/stdlib_shims.cmxa)
File "_build/.dune/default/src/dune", line 2, characters 0-75:
2 | (library
3 |  (wrapped false)
4 |  (name stdlib_shims)
5 |  (public_name stdlib-shims))
Error: Rule failed to generate the following targets:
- src/stdlib_shims.lib
```
Looks like `ocamlopt -a` tries to create an empty library and so there is no .lib file? With OCaml 4.02.3 Win32 I have no such problem.

> However, .lib format doesn't support that (see MSDN - in their infinite wisdom, Microsoft have removed the documentation of LNK1160 from later versions, but it's still the case!).

Actually there is no special "lib" format - it's just the same [archive format](https://en.wikipedia.org/wiki/Ar_(Unix)#File_format_details) that Unix tools like `ar` use. Unfortunately, Microsoft tools not only fail to create empty archives but they also cannot use them. If you manually create a .lib file containing `!<arch>\n` (valid empty archive from `ar` point of view) and try to link it, Microsoft linker will report `fatal error LNK1107: invalid or corrupt file: cannot read at 0x8`

--- a/src/dune
+++ b/src/dune
@@ -58,8 +58,17 @@ let modules_post_4_02 =
   ]
 
 let missing_modules =
-  List.filter modules_post_4_02 ~f:(fun (_, v) -> version < v)
-  |> List.map ~f:fst
+  let missing =
+    List.filter modules_post_4_02 ~f:(fun (_, v) -> version < v)
+    |> List.map ~f:fst
+  in
+    (* For OCaml 4.07-4.09 incl. this solves the problem of being unable to
+       generate empty .cmxa files on MSVC by duplicating the Pervasives module
+       (and updating its deprecation warning not to refer to this library! *)
+    if version < (4, 10) && version >= (4, 7) then
+      "Pervasives"::missing
+    else
+      missing
 
 let all_modules_except_stdlib =
   (modules_in_4_02 @ List.map modules_post_4_02 ~f:fst)

--- a/src/pervasives.ml
+++ b/src/pervasives.ml
@@ -1,0 +1,5 @@
+[@@@ocaml.deprecated "Use Stdlib instead."]
+
+[@@@ocaml.warning "-3"]
+
+include Stdlib.Pervasives

